tests/DCPS/XTypes/Subscriber.cpp

```diff
@@ -4,195 +4,205 @@
+    success = false;
+#else
+    DDS::DynamicData_var use_dd;
+    DDS::MemberId use_id;
+    CORBA::String_var actual;
+    OpenDDS::XTypes::MemberPath member_path;
+    DDS::DynamicType_var dt = ts->get_type();
+    if (!(dd && check_rc(member_path.resolve_string_path(dt, path), "check_string: resolve_string_path") &&
+        check_rc(member_path.get_member_from_data(dd, use_dd, use_id), "check_string: get_member_from_data") &&
+        check_rc(use_dd->get_string_value(actual, use_id), "check_int32: get_int32_value"))) {
+      ACE_ERROR((LM_ERROR, "ERROR: check_string for %C.%C failed\n", type_name.in(), path.c_str()));
+      success = false;
+      return;
     }
-  } else {
-    ACE_ERROR((LM_ERROR, "ERROR: Reader: read_i returned %C\n",
-               OpenDDS::DCPS::retcode_to_string(ret)));
+    check_string(path, expected, actual);
+#endif
   }
 
-  return ret;
-}
+  template <typename DataReaderType, typename SeqType>
+  bool read_single(SeqType& seq)
+  {
+    if (!read_i<DataReaderType>(dr, seq)) {
+      ACE_ERROR((LM_ERROR, "ERROR: read_single %C failed\n", type_name.in()));
+      return false;
+    }
+    if (seq.length() != 1) {
+      ACE_ERROR((LM_ERROR, "ERROR: read %C expected one, but sample got %u\n", seq.length()));
+      return false;
+    }
+    return true;
+  }
 
-template<typename T1, typename T2>
-ReturnCode_t read_extended_struct(const DataReader_var& dr, const T1& pdr, T2& data)
-{
-  ReturnCode_t ret = read_i(dr, pdr, data);
-  if (ret == RETCODE_OK) {
-    if (data.length() != 1) {
-      ACE_ERROR((LM_ERROR, "reader: unexpected data length: %d\n", data.length()));
-      ret = RETCODE_ERROR;
-    } else if (data[0].key_field != key_value) {
-      ACE_ERROR((LM_ERROR, "reader: expected key value: %d, received: %d\n", key_value, data[0].key_field));
-      ret = RETCODE_ERROR;
-    } else if (data[0].additional_field != 0) {
-      ACE_ERROR((LM_ERROR, "reader: expected additional_field: %d, received: %d\n", 0, data[0].additional_field));
-      ret = RETCODE_ERROR;
-    } else if (verbose) {
-      ACE_DEBUG((LM_DEBUG, "reader: %d additional_field: %d\n", data[0].key_field, data[0].additional_field));
+  template <typename SampleType>
+  bool read(SampleType& sample)
+  {
+    typedef typename OpenDDS::DCPS::DDSTraits<SampleType> Traits;
+    typedef typename Traits::MessageSequenceType SeqType;
+    SeqType seq;
+    if (!read_single<typename Traits::DataReaderType, SeqType>(seq)) {
+      return false;
     }
-  } else {
-    ACE_ERROR((LM_ERROR, "ERROR: Reader: read_i returned %C\n",
-               OpenDDS::DCPS::retcode_to_string(ret)));
+    sample = seq[0];
+    return true;
   }
 
-  return ret;
-}
+  bool read_dynamic()
+  {
+#ifdef OPENDDS_SAFETY_PROFILE
+    success = false;
+    return false;
+#else
+    DDS::DynamicDataSeq seq;
+    if (!read_single<DDS::DynamicDataReader, DDS::DynamicDataSeq>(seq)) {
+      return false;
+    }
+    dd = DDS::DynamicData::_duplicate(seq[0]);
```

Comment:
This should also work without the `_duplicate` since the var-to-var assignment rule extends to the unspecified type of sequence elements that's "var-like", but I wouldn't be surprised if that was a gap in the implementation.

Reply:
I don't think I gave any thought when writing it, but removing it causes the subscriber to segfault when trying to read values from the DynamicData, so something's broken. I probably bungled the reference counting somewhere.

Reply:
Okay I've done over my code and it is this assignment because it's not incrementing the reference count without the duplicate.